dsc_lib/src/discovery/command_discovery.rs

```diff
@@ -78,74 +62,241 @@ impl CommandDiscovery {
             }
         }
 
-        for path in paths {
-            if path.exists() && path.is_dir() {
-                for entry in path.read_dir().unwrap() {
-                    let entry = entry.unwrap();
-                    let path = entry.path();
-                    if path.is_file() {
-                        let file_name = path.file_name().unwrap().to_str().unwrap();
-                        let file_name_lowercase = file_name.to_lowercase();
-                        if file_name_lowercase.ends_with(".dsc.resource.json") ||
-                           file_name_lowercase.ends_with(".dsc.resource.yaml") ||
-                           file_name_lowercase.ends_with(".dsc.resource.yml") {
-                            let resource = match load_manifest(&path)
-                            {
-                                Ok(r) => r,
-                                Err(e) => {
-                                    if return_all_resources {
+        // remove duplicate entries to improve perf of resource search
+        let mut uniques = HashSet::new();
+        paths.retain(|e|uniques.insert((*e).clone()));
+
+        Ok(paths)
+    }
+}
+
+impl Default for CommandDiscovery {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ResourceDiscovery for CommandDiscovery {
+
+    #[allow(clippy::too_many_lines)]
+    fn list_available_resources(&mut self, type_name_filter: &str, adapter_name_filter: &str) -> Result<BTreeMap<String, DscResource>, DscError> {
+
+        debug!("Listing resources with type_name_filter/adapter_name_filter: {:?}/{:?}", type_name_filter, adapter_name_filter);
+
+        let pb_span = warn_span!("");
+        pb_span.pb_set_style(&ProgressStyle::with_template(
+            "{spinner:.green} [{elapsed_precise:.cyan}] {msg:.yellow}"
+        )?);
+        pb_span.pb_set_message("Searching for resources");
+        let _ = pb_span.enter();
+
+        let mut resources: BTreeMap<String, DscResource> = BTreeMap::new();
+        let mut adapter_resources: BTreeMap<String, DscResource> = BTreeMap::new();
+
+        let regex_str = convert_wildcard_to_regex(type_name_filter);
+        debug!("Using regex {regex_str} as filter for resource type");
+        let mut regex_builder = RegexBuilder::new(regex_str.as_str());
+        regex_builder.case_insensitive(true);
+        let type_regex: Regex;
+        if let Ok(reg_v) = regex_builder.build() {
+            type_regex = reg_v;
+        } else {
+            let err_str = "Could not build Regex filter for resource type";
+            error!(err_str);
+            return Err(DscError::Operation(err_str.to_string()));
+        }
```

Comment:
I think this could be simplified as

```suggestion
        let Ok(type_regex) = regex_builder.build() else {
            let err_str = "Could not build Regex filter for resource type";
            error!(err_str);
            return Err(DscError::Operation(err_str.to_string()));
        };
```

Reply:
Updated